### PR TITLE
Add fold and biFold functions

### PR DESCRIPTION
**Pull request by chrisdotcode**

### CODE REVIEW

**Comment by chrisdotcode** [comment]
I should probably also update the docs. I'll do that now.

File: src/Data/Path/Pathy.purs [hunk 1]
```diff
@@ -478,6 +477,27 @@ module Data.Path.Pathy
...
+  -- | Folds a `Path` into a value using a specified function and a provided
+  -- | default.
+  fold :: forall c a b s. (String -> c -> c) -> c -> Path a b s -> c
+  fold _ z (Current                 ) = z
+  fold _ z (Root                    ) = z
+  fold f z (ParentIn p              ) = fold f z p
+  fold f z (FileIn   p (FileName f')) = f f' (fold f z p)
+  fold f z (DirIn    p (DirName  f')) = f f' (fold f z p)
+
+  -- | Folds a `Path` into a value using a two specified functions and a
```
Comment (zudov) [inline src/Data/Path/Pathy.purs]: "a two specified functions"

Comment (chrisdotcode) [inline src/Data/Path/Pathy.purs]: Whoops.

**Comment by chrisdotcode** [comment]
@garyb Request to remove `MODULES.md`, and move the docs to the standard purescript doc location? (`docs/`)?

**Comment by garyb** [comment]
Yeah definitely. We could probably switch the build to `pulp` too I think, and then it's just `pulp docs` to generate everything in the appropriate place.

**Comment by chrisdotcode** [comment]
I had actually used both `pulp build` and `pulp docs` just fine to both build and generate docs respectively just now.

File: src/Data/Path/Pathy.purs [hunk 2]
```diff
@@ -478,6 +477,27 @@ module Data.Path.Pathy
...
+  -- | provided default.
+  -- |
+  -- | The first function folds a `FileName` into our value, and the second
+  -- | function folds a `DirName` into our value.
+  biFold :: forall c a b s. (FileName -> c -> c) -> (DirName -> c -> c) -> c -> Path a b s -> c
+  biFold f g z (Current      ) = z
+  biFold f g z (Root         ) = z
+  biFold f g z (ParentIn p   ) = biFold f g z p
+  biFold f g z (FileIn   p f') = f f' (biFold f g z p)
+  biFold f g z (DirIn    p f') = g f' (biFold f g z p)
```
Comment (garyb) [inline src/Data/Path/Pathy.purs]: I don't think `bifold` is the right name here really, as we're not folding a data structure that consists of two parts. I think I'd rather this be just `fold` actually, and do away with the stringly typed version entirely. This is the catamorphism for a `Path`, so `fold` is a perfectly good name for it.

Comment (chrisdotcode) [inline src/Data/Path/Pathy.purs]: :+1:

Comment (garyb) [inline src/Data/Path/Pathy.purs]: There's a `foldJson` in argonaut that works this way, for instance (but with even more inputs).

File: src/Data/Path/Pathy.purs [hunk 3]
```diff
@@ -478,6 +477,27 @@ module Data.Path.Pathy
...
   parseAbsDir = parsePath (const Nothing) (const Nothing) (const Nothing) Just
 
+  -- | Folds a `Path` into a value using a specified function and a provided
+  -- | default.
+  fold :: forall c a b s. (String -> c -> c) -> c -> Path a b s -> c
+  fold _ z (Current                 ) = z
+  fold _ z (Root                    ) = z
+  fold f z (ParentIn p              ) = fold f z p
+  fold f z (FileIn   p (FileName f')) = f f' (fold f z p)
+  fold f z (DirIn    p (DirName  f')) = f f' (fold f z p)
```
Comment (zudov) [inline src/Data/Path/Pathy.purs]: That's a bit nitpickish, but I feel confused when something of type `String` is called `f'`, it probably would be better to call it `a`, but that's up to you I guess.

Comment (chrisdotcode) [inline src/Data/Path/Pathy.purs]: I would have preferred `f` and `d` myself, but I stayed consistent with the rest of the module, where it's `f` everywhere.

If I rename, it'll have to be en masse.

Comment (zudov) [inline src/Data/Path/Pathy.purs]: Oh, I see, sorry for raising it then.

**Comment by chrisdotcode** [comment]
@garyb Before you merge, should I version++, and add git tags accordingly (for bower)?

**Comment by garyb** [comment]
Nah, we usually make the release via GitHub anyway so it does that automatically.

**Comment by chrisdotcode** [comment]
@garyb Ready for merging, then?

**Comment by garyb** [comment]
Could you squash the commits please? And then :+1:

File: src/Data/Path/Pathy.purs [hunk 4]
```diff
@@ -478,6 +476,18 @@ module Data.Path.Pathy
...
 
+  -- | Folds a `Path` into a value using two specified functions and a provided
+  -- | default value.
+  -- |
+  -- | The first function folds a `FileName` into our value, and the second
+  -- | function folds a `DirName` into our value.
+  fold :: forall c a b s. (FileName -> c -> c) -> (DirName -> c -> c) -> c -> Path a b s -> c
+  fold f g z (Current      ) = z
+  fold f g z (Root         ) = z
+  fold f g z (ParentIn p   ) = fold f g z p
```
Comment (jdegoes) [inline src/Data/Path/Pathy.purs:487]: This relationship is interesting and should probably be exposed as part of the fold.

When I was last playing around with adding this, I had something like this:

```
data NamelessNode = ParentPath | CurrentPath | RootPath

-- | Folds over a path, starting at the base of the path and working outward.
foldPath :: forall a b s z. (z -> NamelessNode -> z) -> (z -> Either (Tuple (Path a Dir s) DirName) (Tuple (Path a Dir s) FileName) -> z) -> z -> Path a b s -> z
```

I can speak to how useful this is, but my goal was in _precisely_ capturing the structure of the path (including parent in relationships). In fact, with such a fold function, I wanted to replace the remaining "unsafe" implementations (which switch type parameters around in an unchecked fashion).

Comment (chrisdotcode) [inline src/Data/Path/Pathy.purs:487]: @jdegoes So do you think it's cool to merge for now, and revisit what you had in mind later?

**Comment by chrisdotcode** [comment]
@garyb Do you want one single commit, or do you want me to try and preserve the code changes/documentation changes if and where possible?

**Comment by garyb** [comment]
We tend to just squash into one commit unless the change is huge :wink:

I'll leave the final merging to @jdegoes though, given he has ideas about how the fold should work.

**Comment by jdegoes** [comment]
I feel this version of `fold` is easy to misuse: for example, if I try to construct a path string by folding over the path, then the resulting path will be inaccurate because it fails to take into account `../foo/../bar/` (`ParentIn`) nodes. I would be open to (a) fixing this issue, or (b) renaming it `foldNames` and making sure the documentation is clear it will fold only over the file and dir names, not all the nodes making up the path.

**Comment by chrisdotcode** [comment]
@jdegoes How about exposing the constructors in some way, shape or form, like in your example above?

**Comment by jdegoes** [comment]
Thanks so much for this @chrisdotcode!

**Comment by jdegoes** [comment]
Oops, accidentally merged this prior to the final revision! Sorry. :disappointed:

**Pull request merged**